### PR TITLE
meta(devtools): initial direnv configuration

### PR DESCRIPTION
**Pull request by joshuarli**
Here's an initial envrc for direnv, as part of an effort to standardize/enforce the sentry dev environment.

It sets some useful environment variables, and does reasonably quick (time budget TBD) state checking. If initialization is needed to bring up state, like creating `.venv/`, it will do so. Or, in some cases it'll recommend to run certain commands.

### CODE REVIEW

**Comment by dcramer** [comment]
I would add support for loading a .env file to allow folks to overload config. This is something I know @mattrobenolt mentioned he wanted, and its a good way to store secrets (throw it into .gitignore!) or test changes.

**Comment by mattrobenolt** [comment]
Yeah, the only thing I care about here is I already heavily use direnv and have a local .envrc in a lot of repositories that I depend on with my own personal things. I would really like to preserve the ability to set my own stuff, even if it means I have to move to a `.envrc-extras` or something file.

**Comment by dcramer** [comment]
@mattrobenolt do you only set env vars or do you also execute code?

File: .envrc [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+    >&2 cat <<EOF
+${red}${bold}direnv FATAL: ${1}
+${reset}
+EOF
+    exit 1
+}
+
+init_venv () {
+    deactivate 2>/dev/null || true
+    python2.7 -c "import sys; sys.exit(sys.version_info[:2] != (2, 7))" || \
```
Comment (mattrobenolt) [inline .envrc]: Why would you even need this if you're explicitly calling `python2.7` binary? I mean, I agree that that doesn't guarantee it, but you'd have to go out of your way to do something really odd here like symlink python3 to python2.7 or something.

File: .envrc [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+        sentry_redis        ;
+        do
+        docker exec "$container" true || die "The docker container ${container} doesn't seem to be running.
+    Please run sentry devservices up."
+    done
+fi
+
+# TODO: is there a faster way to state-check the need for sentry reset-db and sentry createuser?
+# This is a little slower than i'd like, so not turning it on for now, not even behind a NO_CHECK flag.
+# docker exec sentry_postgres sh -c "psql -U postgres -h 127.0.0.1 sentry --command 'select * from sentry_useremail'"
```
Comment (mattrobenolt) [inline .envrc]: Slightly faster query would just just be `select 1 from sentry_useremail`. I'm not sure if that's the source of it being slow or not though.

**Comment by mattrobenolt** [comment]
@dcramer pretty sure I only set env vars. At least in sentry+getsentry.

File: .envrc [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+#   - lightweight (crucial since it's run on every cd to here) state checking and assurance (e.g. activating the venv)
+#   - one-time heavyweight state initialization (such as creating a .venv)
+#       - TODO: heavyweight state assurance, such as syncing venv, should happen in a one-off "fixall" script
+#               and should be recommended by direnv as a catch-all fixer
+#       - Although, this kind of state initialization could be moved to there from here in the future.
+#   - setting useful environment variables
+
+export PYTHONDONTWRITEBYTECODE=1
+export PIP_DISABLE_PIP_VERSION_CHECK=on
+export NODE_OPTIONS=--max-old-space-size=4096
```
Comment (dcramer) [inline .envrc]: any chance we could add comments to what these do

File: .envrc [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+    do
+    if ! require "$pkg"; then
+        die "You seem to be missing the system dependency: ${pkg}
+Please install homebrew, and run brew bundle."
+    fi
+done
+
+### Python ###
+
+# direnv set -u's; unsure if it's a good idea to set +u.
```
Comment (dcramer) [inline .envrc]: i personally am a fan of -u as it prevents some errors

Comment (joshuarli) [reply .envrc]: Yeah, which is why i'm doing stuff like `${VIRTUAL_ENV:-}` as needed.

File: .envrc (hunk 2 above)
Comment (joshuarli) [reply .envrc]: With my dev database, there's only 1 user - speed is roughly the same haha. Thanks though.

File: .envrc (hunk 1 above)
Comment (joshuarli) [reply .envrc]: Good point. Admittedly this was copypaste from one of my old branches where I didn't have a following `python2.7 -m virtualenv .venv`. I'll replace it with a command check.

File: .envrc [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+reset="$(tput sgr0)"
+
+# NOTE: do not trap bash's EXIT; it will prevent direnv from performing its final dance(s)
+trap notice ERR
+
+notice () {
+    [ $? -eq 0 ] && return
+    cat <<EOF
+${bold}direnv tooling is in an ALPHA state!
+If you're having trouble, or have questions, please ask in #discuss-dev-tooling
```
Comment (markstory) [inline .envrc]: Is this a channel that is coming soon?

Comment (joshuarli) [reply .envrc]: It's public, but I added you.

File: .envrc [hunk 6]
```diff
@@ -0,0 +1,195 @@
+# This is the .envrc for Sentry, for use with direnv.
+# It's responsible for enforcing a standard dev environment by checking as much state as possible, and either performing
+# initialization (e.g. activating the venv) or giving recommendations on how to reach the desired state.
+# It also sets useful environment variables.
+# If you'd like to override or do any custom things, this .envrc will source a .envrc-overrides file at the end.
```
Comment (EvanPurkhiser) [inline .envrc]: Total bike shed, but why not wrap this to 80 col

Comment (joshuarli) [reply .envrc]: 🙄

File: .envrc [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+${red}${bold}direnv FATAL: ${1}
+${reset}
+EOF
+    return 1
+}
+
+init_venv () {
+    deactivate 2>/dev/null || true
+    info "Creating a virtualenv for you."
+    require python2.7 || die "You'll need to install python2.7, or make it available on your PATH. It's recommended to use pyenv."
```
Comment (EvanPurkhiser) [inline .envrc]: If we're going to enforce pyenv, can we just say "Use `pyenv install`"

File: .envrc [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+info "Checking node..."
+
+node_version="10.16.3"
+
+# It would be nice to enforce that node is installed via volta (and is therefore a shim that will check against
+# the node pin in package.json), but for now, let's just explicitly check the node version.
+
+if ! require node; then
+    die "You don't seem to have node installed. We want version ${node_version}.
+It's recommended to use https://github.com/volta-cli/volta."
```
Comment (EvanPurkhiser) [inline .envrc]: When you say above `recommended to use pyenv` you don't link anything, can we be consistent and either link pyenv or remove this link.

Comment (joshuarli) [reply .envrc]: What do you think about 2057752c32224b731c4e423d907c0e35efe39396?

File: .envrc [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+    info "You don't seem to have yarn packages installed. Let's install them."
+    make install-yarn-pkgs
+fi
+
+PATH_add node_modules/.bin
+
+
+### Overrides ###
+
+if [ -f .envrc-overrides ]; then
```
Comment (dcramer) [inline .envrc]: This is a risk fwiw. If someone writes to this file it’ll get run without confirmation. That’s why I chose to do .env only.

**Review by dcramer — APPROVED** [review]

File: .envrc (hunk 9 above)
Comment (joshuarli) [reply .envrc]: Sure. It's for power users though, I think it's fair to assume they know what they're doing.

Comment (dcramer) [reply .envrc]: I’m saying I can root your machine. Doesn’t matter if you’re a power user.

Comment (dcramer) [reply .envrc]: I’m saying I can root your machine. Doesn’t matter if you’re a power user.

Comment (joshuarli) [reply .envrc]: Discussed offline and I was convinced. 16999de

**Pull request merged**